hazelcast-jet-sql/src/main/java/com/hazelcast/jet/sql/impl/aggregate/AvgSqlAggregation.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jet.sql.impl.aggregate;
+
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.sql.impl.QueryException;
+import com.hazelcast.sql.impl.type.QueryDataType;
+import com.hazelcast.sql.impl.type.QueryDataTypeFamily;
+
+import javax.annotation.concurrent.NotThreadSafe;
+import java.io.IOException;
+import java.math.BigDecimal;
+import java.util.Objects;
+
+import static com.hazelcast.sql.impl.expression.math.ExpressionMath.DECIMAL_MATH_CONTEXT;
+
+@NotThreadSafe
+public class AvgSqlAggregation extends SqlAggregation {
+
+    private QueryDataType resultType;
+
+    private SumSqlAggregation sum;
+    private CountSqlAggregation count;
+
+    @SuppressWarnings("unused")
+    private AvgSqlAggregation() {
+    }
+
+    public AvgSqlAggregation(int index, QueryDataType operandType) {
+        this(index, operandType, false);
+    }
+
+    public AvgSqlAggregation(int index, QueryDataType operandType, boolean distinct) {
+        super(index, true, distinct);
+        this.resultType = inferResultType(operandType);
+        this.sum = new SumSqlAggregation(index, operandType);
+        this.count = new CountSqlAggregation();
+    }
+
+    private static QueryDataType inferResultType(QueryDataType operandType) {
+        switch (operandType.getTypeFamily()) {
+            case TINYINT:
+            case SMALLINT:
+            case INTEGER:
+            case BIGINT:
+            case DECIMAL:
```

Comment:
Converging the type to DECIMAL aggressively may have bad performance consequences, working with BigDecimal is resource-intensive. For IMDG we generally do not extend the type to `DECIMAL`. E.g. `BIGINT+BIGINT=>BIGINT`, not DECIMAL. This brings better performance, and in the case of overflow, we may always ask the user to add an explicit CAST. When it is time to implement aggregations in IMDG, we are likely to follow the same approach. E.g. `SUM(INTEGER)=>BIGINT`, but `SUM(BIGINT)=>BIGINT`.

I would suggest you considering less aggressive type widening rules, to avoid `BigDecimal` as much as possible.

Reply:
For sums we have `BIGINT+BIGINT=>BIGINT`. How would you represent a fraction with `BIGINT` for averages?

Reply:
In postgres it also returns a decimal likely, certainly not an `int`: https://dbfiddle.uk/?rdbms=postgres_13&fiddle=af9ef053f45de269a71cf41ab8fdf9fb

Reply:
Ah, sorry. So `BigDecimal` appears at the very end, and sum/count is performed without decimals for most type